ogre2/src/Ogre2Projector.cc

```diff
@@ -0,0 +1,339 @@
+#include <OgreSceneNode.h>
+#include <OgreTextureGpuManager.h>
+
+#include "gz/rendering/ogre2/Ogre2Camera.hh"
+#include "gz/rendering/ogre2/Ogre2DepthCamera.hh"
+#include "gz/rendering/ogre2/Ogre2Projector.hh"
+#include "gz/rendering/ogre2/Ogre2RenderEngine.hh"
+#include "gz/rendering/ogre2/Ogre2Scene.hh"
+#include "gz/rendering/Utils.hh"
+
+namespace gz
+{
+namespace rendering
+{
+inline namespace GZ_RENDERING_VERSION_NAMESPACE {
+//
+/// \brief Helper class for checking visibility of projecter to a camera
+class Ogre2ProjectorCameraListener: public Ogre::Camera::Listener
+{
+  /// \brief Constructor
+  /// \param[in] _decal Pointer to the ogre decal (projected texture)
+  public: Ogre2ProjectorCameraListener(Ogre::Decal *_decal);
+
+  //// \brief Set the visibility flags for this projector
+  /// \param[in] _flags Visibility flags to set
+  public: void SetVisibilityFlags(uint32_t _flags);
+
+  /// \brief Callback when a camara is about to be rendered
+  /// \param[in] _cam Ogre camera pointer which is about to render
+  private: virtual void cameraPreRenderScene(
+    Ogre::Camera * _cam) override;
+
+  /// \brief Callback when a camera is finisned being rendered
+  /// \param[in] _cam Ogre camera pointer which has already render
+  private: virtual void cameraPostRenderScene(
+    Ogre::Camera * _cam) override;
+
+  /// \brief Projector's visibility flags
+  private: uint32_t visibilityFlags = 0u;
+
+  /// \brief Pointer to the decal ogre scene node
+  public: Ogre::SceneNode *decalNode{nullptr};
+
+  /// \brief Decal - Texture projected onto a surface
+  public: Ogre::Decal *decal{nullptr};
+};
+}
+}
+}
+
+using namespace gz;
+using namespace rendering;
+
+/// \brief Private data for the Ogre2Projector class
+class gz::rendering::Ogre2Projector::Implementation
+{
+  /// \brief The decal ogre scene node
+  public: Ogre::SceneNode *decalNode{nullptr};
+
+  /// \brief Decal diffuse texture
+  public: Ogre::TextureGpu *textureDiff{nullptr};
+
+  /// \brief Decal - Texture projected onto a surface
+  public: Ogre::Decal *decal{nullptr};
+
+  /// \brief Indicate whether the projector is intialized or not
+  public: bool initialized{false};
+
+  /// \brief A map of cameras (<Camera ptr, name>) that the listener has been
+  /// added to
+  public: std::unordered_map<Ogre::Camera *, Ogre::IdString>
+      camerasWithListener;
+
+  /// \brief Listener for togging projector visibility
+  /// We are using a custom listener because Ogre::Decal's setVisibilityFlags
+  /// does not seem to work
+  public: std::unique_ptr<Ogre2ProjectorCameraListener> listener;
+};
+
+/////////////////////////////////////////////////
+Ogre2Projector::Ogre2Projector()
+    : dataPtr(utils::MakeUniqueImpl<Implementation>())
+{
+}
+
+/////////////////////////////////////////////////
+Ogre2Projector::~Ogre2Projector()
+{
+  this->SetEnabled(false);
+
+  if (!this->scene->IsInitialized())
+    return;
+
+  for (auto &ogreCamIt : this->dataPtr->camerasWithListener)
```

Comment:
Should `ogreCamIt` be `const`?

Reply:
added `const`. [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)